feat: rollback new site creation on failure

akhilnarang opened the pull request:
- chore: drop unused parameter
- chore: update output to let the user know they can pass `--force`
- feat: generate a non-deterministic database name every time - include current date in the hash
- feat: initial failed site rollback implementation

Resolves #25671

Docs: https://frappeframework.com/docs/user/en/tutorial/create-a-site

akhilnarang committed:
chore: drop unused parameter

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: update output to let the user know they can pass `--force`

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: generate a non-deterministic database name every time - include current date in the hash

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: initial failed site rollback implementation

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/installer.py:
```diff
@@ -123,6 +127,22 @@ def _new_site(
+	db_user=None,
+	setup_db=True,
+):
+	if frappe.flags.db_name:
+		if setup_db:
+			from frappe.database import drop_user_and_database
+
+			drop_user_and_database(frappe.flags.db_name, db_user or frappe.flags.db_name)
+
+		if os.path.isdir(frappe.get_site_path()):
+			import shutil
+
+			shutil.rmtree(frappe.get_site_path())
```
IMO `CallbackManager` would be better choice here for rolling back only the changes that have happened. It will also setup a pattern for what needs to be done for any rollback-able action.

akhilnarang committed:
chore: check whether we're in an interactive session or not

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: check if rollback callback is defined

Don't raise exceptions where a simple print() will suffice
Print exception before prompting for rollback

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: switch to `datetime.now()` instead of our `get_datetime()`

`get_datetime()` will query the DB to get the system timezone, not something we currently have.

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: pass function as lambda

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/commands/site.py:
```diff
@@ -70,29 +72,39 @@ def new_site(
-		use(site)
+	except Exception:
+		traceback.print_exc()
+		if sys.__stdin__.isatty() and click.confirm("Do you want to rollback the failed site?", abort=True):
```
We should clarify that site creation has failed in this message.

ankush replied:
I hope FC isn't faking a TTY, else it will just stay stuck forever on prompt :smile:

ankush commented on frappe/installer.py:
```diff
@@ -70,11 +71,13 @@ def _new_site(
 			"_"
 			+ hashlib.sha1(
-				os.path.realpath(frappe.get_site_path()).encode(), usedforsecurity=False
+				(os.path.realpath(frappe.get_site_path()) + datetime.now().isoformat()).encode(),
+				usedforsecurity=False,
 			).hexdigest()[:16]
```
This is `frappe.generate_hash(length=16)` with extra steps :smile:

akhilnarang committed:
fix(generate_hash): Deprecate `txt` parameter - its not used.

`length` has a default value anyway, no need of checking and setting a default.

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(installer): use `frappe.generate_hash()` to generate database name

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(new-site): update rollback prompt

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang commented:
@mergifyio backport version-15-hotfix version-14-hotfix
